File: src/carpcomm/scheduler/pass_details.go
// Author: Timothy Stranex <tstranex@carpcomm.com>
// Copyright 2013 Timothy Stranex

package scheduler

import "os/exec"
import "fmt"
import "time"

type SatPoint struct {
	Timestamp float64
	AzimuthDegrees float64
	AltitudeDegrees float64
	Range float64  // [m]
	RangeVelocity float64  // [m/s]
	LatitudeDegrees float64
	LongitudeDegrees float64
	Elevation float64  // [m]
	IsEclipsed bool
}


const passDetailsBinaryPath = "src/carpcomm/scheduler/pass_details.py"

func passDetails(
	binary_path string,
	begin_time time.Time,
	duration time.Duration,
	latitude_degrees,
	longitude_degrees,
	elevation_metres float64,
	tle string,
	resolution_seconds float64) ([]SatPoint, error) {

	begin_timestamp := 1e-9 * float64(begin_time.UnixNano())

	c := exec.Command("python", binary_path)
	in_pipe, err := c.StdinPipe()
	if err != nil {
		return nil, err
	}
	defer in_pipe.Close()
	out_pipe, err := c.StdoutPipe()
	if err != nil {
		return nil, err
	}
	defer out_pipe.Close()
	if err := c.Start(); err != nil {
		return nil, err
	}
	defer c.Wait()

	fmt.Fprintf(in_pipe, "%f\n%f\n%f\n%f\n%f\n%s\n%f\n",
		begin_timestamp,
		duration.Seconds(),
		latitude_degrees,
		longitude_degrees,
		elevation_metres,
		tle,
		resolution_seconds)
	in_pipe.Close()

	var n int
	_, err = fmt.Fscanln(out_pipe, &n)
	if err != nil {
		return nil, err
	}

	r := make([]SatPoint, n)
	for i := 0; i < n; i++ {
		p := &r[i]
		fmt.Fscanln(out_pipe,
			&p.Timestamp,
			&p.AzimuthDegrees,
			&p.AltitudeDegrees,
			&p.Range,
			&p.RangeVelocity,
			&p.LatitudeDegrees,
			&p.LongitudeDegrees,
			&p.Elevation,
			&p.IsEclipsed)
	}
	return r, nil
}

func PassDetails(begin_time time.Time,
	duration time.Duration,
	latitude_degrees,
	longitude_degrees,
	elevation_metres float64,
	tle string,
	resolution_seconds float64) ([]SatPoint, error) {

	return passDetails(passDetailsBinaryPath,
		begin_time,
		duration,
		latitude_degrees,
		longitude_degrees,
		elevation_metres,
		tle,
		resolution_seconds)
}

File: src/carpcomm/scheduler/pass_details.py
#!/usr/bin/python

# Author: Timothy Stranex <tstranex@carpcomm.com>
# Copyright 2013 Timothy Stranex

#!/usr/bin/python

"""
Example input:

1349814979.0
120.0
47.4
8.5
400.0
1998-067CQ
1 38854U 98067CP  12283.07336473  .00054398  00000-0  89879-3 0    14
2 38854  51.6473 275.4897 0014651 145.1372 215.0744 15.51582271   671
1.0
"""

import ephem
import math
import sys
import datetime

def PassDetails(begin_timestamp, duration_seconds,
                latitude_degrees, longitude_degrees, elevation_metres,
                tle, resolution_seconds):
    obs = ephem.Observer()
    obs.lat = math.radians(latitude_degrees)
    obs.lon = math.radians(longitude_degrees)
    obs.elevation = elevation_metres

    body = ephem.readtle(*tle)

    begin_time = ephem.Date(datetime.datetime.utcfromtimestamp(begin_timestamp))

    n = int(duration_seconds/resolution_seconds)
    print n
    for i in xrange(n):
        t_offset = i*resolution_seconds
        timestamp = begin_timestamp + t_offset
        obs.date = ephem.Date(begin_time + ephem.second*t_offset)
        body.compute(obs)

        print '%f %f %f %f %f %f %f %f %d' % (
            timestamp,
            math.degrees(body.az),
            math.degrees(body.alt),
            float(body.range), # [m]
            float(body.range_velocity), # [m/s]
            math.degrees(body.sublat),
            math.degrees(body.sublong),
            float(body.elevation), # [m]
            bool(body.eclipsed))


def main():
    PassDetails(float(sys.stdin.readline()),
                float(sys.stdin.readline()),
                float(sys.stdin.readline()),
                float(sys.stdin.readline()),
                float(sys.stdin.readline()),
                [sys.stdin.readline(),
                 sys.stdin.readline(),
                 sys.stdin.readline()],
                float(sys.stdin.readline()))
                

if __name__ == '__main__':
    main()
